library/std/src/diagnostics.qs

```diff
@@ -69,6 +69,44 @@ namespace Microsoft.Quantum.Diagnostics {
         body intrinsic;
     }
 
+    /// # Summary
+    /// Given an operation, dumps the matrix representation of the operation actiong on the given
+    /// number of qubits.
+    ///
+    /// # Input
+    /// ## nQubits
+    /// The number of qubits on which the given operation acts.
+    /// ## op
+    /// The operation that is to be diagnosed.
+    ///
+    /// # Remarks
+    /// When run on the sparse-state simulator, the following snippet
+    /// will output the matrix
+    /// $\left(\begin{matrix} 0.0 & 0.707 \\\\ 0.707 & 0.0\end{matrix}\right)$:
+    ///
+    /// ```qsharp
+    /// operation DumpH() : Unit {
+    ///     DumpOperation(1, qs => H(qs[0]));
+    /// }
+    /// ```
+    /// Calling this operation has no observable effect from within Q#.
+    /// Note that if `DumpOperation` is called when there are other qubits allocated,
+    /// the matrix displayed may reflect any global phase that has accumulated on the other qubits.
+    operation DumpOperation(nQubits : Int, op : Qubit[] => Unit) : Unit {
```

Comment:
I'd prefer to not make it generic on the return, as operations that return measurement results tend to not be representable as a matrix. A quantum operation is "unitary" when it can be represented by a unitary matrix, which is not the same thing as returning `Unit` but we sometimes conflate. This is because a unitary operation is reversible, so it has a defined (or generated) adjoint, and only operations that return `Unit` can have an adjoint. In fact, if you look at the classic QDK definition of `DumpOperation` [over here](https://github.com/microsoft/QuantumLibraries/blob/f2f5e380225367ee4eb995d922067025dc77f0a0/Standard/src/Diagnostics/Dump.qs#L71) it actually required in the signature that the provided operation was adjointable even though it didn't use the adjoint. Thinking about this in answer to your question, I realized my intent to relax the restriction of adjointable may have worked against its subtle intent here: to ensure that the operation is something that can be represented as a unitary. So I'm starting to think that rather than relaxing further with a generic return, this should actually be as strict as the original and require `Qubit[] => Unit is Adj` in the type.

Reply:
I did end up making this more restrictive, so now it requires an adjointable operation. I'll need to circle back and make the same change to the Python version for consistency.